packages/veritone-widgets/src/redux/modules/mediaDetails/index.js

```diff
@@ -0,0 +1,657 @@
+import {
+  get,
+  has,
+  findLastIndex,
+  findIndex,
+  groupBy,
+  forEach,
+  map,
+  values,
+  uniqBy
+} from 'lodash';
+import { helpers } from 'veritone-redux-common';
+const { createReducer } = helpers;
+
+export const LOAD_ENGINE_CATEGORIES_COMPLETE =
+  'LOAD_ENGINE_CATEGORIES_COMPLETE';
+export const LOAD_ENGINE_RESULTS = 'LOAD_ENGINE_RESULTS';
+export const LOAD_ENGINE_RESULTS_COMPLETE = 'LOAD_ENGINE_RESULTS_COMPLETE';
+export const LOAD_TDO = 'LOAD_TDO';
+export const LOAD_TDO_SUCCESS = 'LOAD_TDO_SUCCESS';
+export const UPDATE_TDO = 'UPDATE_TDO';
+export const UPDATE_TDO_COMPLETE = 'UPDATE_TDO_COMPLETE';
+export const LOAD_CONTENT_TEMPLATES = 'LOAD_CONTENT_TEMPLATES';
+export const LOAD_CONTENT_TEMPLATES_COMPLETE =
+  'LOAD_CONTENT_TEMPLATES_COMPLETE';
+export const LOAD_TDO_CONTENT_TEMPLATES_COMPLETE =
+  'LOAD_TDO_CONTENT_TEMPLATES_COMPLETE';
+export const UPDATE_TDO_CONTENT_TEMPLATES = 'UPDATE_TDO_CONTENT_TEMPLATES';
+export const UPDATE_TDO_CONTENT_TEMPLATES_FAILURE =
+  'UPDATE_TDO_CONTENT_TEMPLATES_FAILURE';
+export const SELECT_ENGINE_CATEGORY = 'SELECT_ENGINE_CATEGORY';
+export const SET_SELECTED_ENGINE_ID = 'SET_SELECTED_ENGINE_ID';
+export const TOGGLE_EDIT_MODE = 'TOGGLE_EDIT_MODE';
+export const TOGGLE_INFO_PANEL = 'TOGGLE_INFO_PANEL';
+export const INITIALIZE_WIDGET = 'INITIALIZE_WIDGET';
+export const ADD_ENGINE_RESULTS_REQUEST = 'ADD_ENGINE_RESULTS_REQUEST';
+export const TOGGLE_EXPANDED_MODE = 'TOGGLE_EXPANDED_MODE';
+export const REQUEST_LIBRARIES = 'REQUEST_LIBRARIES';
+export const REQUEST_LIBRARIES_SUCCESS = 'REQUEST_LIBRARIES_SUCCESS';
+export const REQUEST_LIBRARIES_FAILURE = 'REQUEST_LIBRARIES_FAILURE';
+export const REQUEST_ENTITIES = 'REQUEST_ENTITIES';
+export const REQUEST_ENTITIES_SUCCESS = 'REQUEST_ENTITIES_SUCCESS';
+export const REQUEST_ENTITIES_FAILURE = 'REQUEST_ENTITIES_FAILURE';
+export const REQUEST_SCHEMAS = 'REQUEST_SCHEMAS';
+export const REQUEST_SCHEMAS_SUCCESS = 'REQUEST_SCHEMAS_SUCCESS';
+export const REQUEST_SCHEMAS_FAILURE = 'REQUEST_SCHEMAS_FAILURE';
+
+export const namespace = 'mediaDetails';
+
+const defaultMDPState = {
+  engineCategories: [],
+  engineResultsByEngineId: {},
+  engineResultRequestsByEngineId: {}, // A list of engine result request so we don't request already fetched data.
```

Comment:
good thought but we should optimize this to avoid duplicating if possible

Reply:
done